File: internal/computer/approval_policy.go
package computer

import (
	"regexp"
	"strings"
)

type ApprovalMode string

const (
	ApprovalAskEveryTime  ApprovalMode = "ask_every_time"
	ApprovalScopedChanges ApprovalMode = "scoped_changes"
	ApprovalFullTask      ApprovalMode = "full_task"
)

type ActionClass string

const (
	ActionReadOnly      ActionClass = "read_only"
	ActionReversible    ActionClass = "reversible"
	ActionConsequential ActionClass = "consequential"
	ActionForbidden     ActionClass = "forbidden"
)

func (class ActionClass) Valid() bool {
	switch class {
	case ActionReadOnly, ActionReversible, ActionConsequential, ActionForbidden:
		return true
	}
	return false
}

func (mode ApprovalMode) Valid() bool {
	switch mode {
	case ApprovalAskEveryTime, ApprovalScopedChanges, ApprovalFullTask:
		return true
	}
	return false
}

func (mode ApprovalMode) Allows(class ActionClass) bool {
	if !mode.Valid() || class == ActionForbidden {
		return false
	}
	switch mode {
	case ApprovalScopedChanges:
		return class == ActionReadOnly || class == ActionReversible
	case ApprovalFullTask:
		return class != ActionForbidden
	default:
		return class == ActionReadOnly
	}
}

var forbiddenControl = regexp.MustCompile(`(?i)\b(password|passcode|credential|secret|token|one[ -]?time|verification code|credit card|card number|cvv|cvc|payment|pay now|purchase|buy now|checkout|bank|wire|transfer money|install|update software|administrator|admin access|security setting|firewall|antivirus|permanent(?:ly)? delete|delete forever|erase)\b`)
var reversibleControl = regexp.MustCompile(`(?i)\b(save draft|save|apply|add|create draft|update draft)\b`)

// ClassifyOperation is deliberately conservative. Labels and roles come from
// the local platform driver, never from the model or service request.
func ClassifyOperation(op Operation, label, role string) ActionClass {
	context := strings.TrimSpace(label + " " + role)
	if forbiddenControl.MatchString(context) {
		return ActionForbidden
	}
	switch op.Kind {
	case "observe", "capture", "focus":
		return ActionReadOnly
	case "replace_text", "select", "set_checked", "scroll":
		return ActionReversible
	case "shortcut":
		switch op.Shortcut {
		case "copy", "undo", "redo", "select_all", "save", "escape", "tab", "reverse_tab", "left", "right", "up", "down", "page_up", "page_down", "home", "end", "paste":
			return ActionReversible
		default:
			return ActionConsequential
		}
	case "activate":
		if reversibleControl.MatchString(context) {
			return ActionReversible
		}
		return ActionConsequential
	case "navigate":
		return ActionReversible
	case "upload", "download", "create_file", "copy_file", "rename_file", "move_file", "trash_file", "overwrite_file", "coordinate_activate":
		return ActionConsequential
	default:
		return ActionForbidden
	}
}

File: computer/approval-policy.mjs
export const APPROVAL_MODES = Object.freeze(['ask_every_time','scoped_changes','full_task']);
export const ACTION_CLASSES = Object.freeze(['read_only','reversible','consequential','forbidden']);

export function validApprovalMode(value) { return APPROVAL_MODES.includes(value); }
export function validActionClass(value) { return ACTION_CLASSES.includes(value); }
export function policyAllows(mode, actionClass) {
  if (!validApprovalMode(mode) || !validActionClass(actionClass) || actionClass === 'forbidden') return false;
  if (mode === 'full_task') return true;
  if (mode === 'scoped_changes') return actionClass === 'read_only' || actionClass === 'reversible';
  return actionClass === 'read_only';
}

const blocked = /\b(password|passcode|credential|secret|token|one[ -]?time|verification code|credit card|card number|cvv|cvc|payment|pay now|purchase|buy now|checkout|bank|wire|transfer money|install|update software|administrator|admin access|security setting|firewall|antivirus|permanent(?:ly)? delete|delete forever|erase)\b/i;
const reversible = /\b(save draft|save|apply|add|create draft|update draft)\b/i;

export function classifyBrowserAction(kind, control = {}) {
  const context = `${control.label ?? ''} ${control.type ?? ''} ${control.tag ?? ''} ${control.riskText ?? ''}`;
  if (blocked.test(context)) return 'forbidden';
  if (['browser_observe','browser_capture','browser_verify'].includes(kind)) return 'read_only';
  if (['browser_navigate','browser_fill','browser_select','browser_set_checked'].includes(kind)) return 'reversible';
  if (['browser_download','browser_upload'].includes(kind)) return 'consequential';
  if (kind === 'browser_click') return reversible.test(context) ? 'reversible' : 'consequential';
  return 'forbidden';
}
